### PR TITLE
[promptflow][internal] Support trace UI page

### PR DESCRIPTION
**Pull request by zhengfeiwang**
# Description

This PR targets to support real trace UI page:

image

**Trace related APIs**

- Spans: `/v1.0/Spans/list`
- LineRuns: `/v1.0/LineRuns/list`

Both APIs have strong type in Swagger models, however, for `attributes` and `resource.attributes`, we still use raw dictionary as marshmallow (flask-restx back) cannot parse field name with dot (".").

**Other changes**

- Expose `start_trace`, so that user can use `pf.start_trace` or `from promptflow import start_trace` - not announce this until we are fully ready.
- Convert the type of `LineRun` evaluations from dictionary to list of dictionary to better support strong type.
- Support CORS handling for PFS, especially for UX local development.

# All Promptflow Contribution checklist:
- [x] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_sdk/_service/apis/line_run.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        LineRunFieldName.INPUTS: fields.Raw(required=True),
+        LineRunFieldName.OUTPUTS: fields.Raw(required=True),
+        LineRunFieldName.START_TIME: fields.DateTime(required=True, dt_format=PFS_MODEL_DATETIME_FORMAT),
+        LineRunFieldName.END_TIME: fields.DateTime(required=True, dt_format=PFS_MODEL_DATETIME_FORMAT),
+        LineRunFieldName.STATUS: fields.String(required=True),
+        LineRunFieldName.LATENCY: fields.String(required=True),
+        LineRunFieldName.NAME: fields.String(required=True),
+        LineRunFieldName.KIND: fields.String(required=True),
+        LineRunFieldName.CUMULATIVE_TOKEN_COUNT: fields.String,
+        LineRunFieldName.EVALUATIONS: fields.Raw,
```
Comment (zhengfeiwang) [inline src/promptflow/promptflow/_sdk/_service/apis/line_run.py:67]: `<eval-name>: {}` is not REST, so might not be able to express with strong type

File: src/promptflow/promptflow/_sdk/_constants.py [hunk 2]
```diff
@@ -441,3 +441,34 @@ class EnvironmentVariables:
     """The environment variables."""
 
     PF_USE_AZURE_CLI_CREDENTIAL = "PF_USE_AZURE_CLI_CREDENTIAL"
+
+
+class EvaluationLineRunFieldName:
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_constants.py:446]: shall we reuse the LineRunFieldName class?

File: src/promptflow/promptflow/_sdk/_service/apis/line_run.py (hunk 1 above)
Comment (zhengfeiwang) [reply src/promptflow/promptflow/_sdk/_service/apis/line_run.py]: updated to the list of dict format

File: src/promptflow/promptflow/_sdk/entities/_trace.py [hunk 3]
```diff
@@ -258,9 +259,14 @@ def _from_spans(spans: typing.List[Span]) -> typing.Optional["LineRun"]:
         if main_line_run_data is None:
             return None
 
-        evaluations = dict()
+        evaluations = []
         for eval_name, eval_line_run_data in evaluation_line_run_data_dict.items():
-            evaluations[eval_name] = eval_line_run_data
+            evaluations.append(
+                {
+                    EvaluationKeyName.NAME: eval_name,
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/entities/_trace.py:266]: why not just: evaluations.append(eval_line_run_data)
eval_line_run_data should have the eval_name already?

File: src/promptflow/promptflow/_sdk/_service/apis/line_run.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        LineRunFieldName.LINE_RUN_ID: fields.String(required=True),
+        LineRunFieldName.TRACE_ID: fields.String(required=True),
+        LineRunFieldName.ROOT_SPAN_ID: fields.String(required=True),
+        LineRunFieldName.INPUTS: fields.Raw(required=True),
+        LineRunFieldName.OUTPUTS: fields.Raw(required=True),
+        LineRunFieldName.START_TIME: fields.DateTime(required=True, dt_format=PFS_MODEL_DATETIME_FORMAT),
+        LineRunFieldName.END_TIME: fields.DateTime(required=True, dt_format=PFS_MODEL_DATETIME_FORMAT),
+        LineRunFieldName.STATUS: fields.String(required=True),
+        LineRunFieldName.LATENCY: fields.String(required=True),
+        LineRunFieldName.NAME: fields.String(required=True),
```
[the remark this answers is not in the document]
Comment (zhengfeiwang) [reply src/promptflow/promptflow/_sdk/_service/apis/line_run.py]: updated

File: src/promptflow/promptflow/_sdk/entities/_trace.py (hunk 3 above)
Comment (zhengfeiwang) [reply src/promptflow/promptflow/_sdk/entities/_trace.py]: yes, just synced with Kaiyi, we should avoid that name

File: src/promptflow/promptflow/_sdk/entities/_trace.py [hunk 5]
```diff
@@ -246,7 +247,12 @@ def _from_spans(spans: typing.List[Span]) -> typing.Optional["LineRun"]:
                 SpanAttributeFieldName.REFERENCED_LINE_RUN_ID in attributes  # test scenario
                 or SpanAttributeFieldName.REFERENCED_BATCH_RUN_ID in attributes  # batch run scenario
             ):
-                evaluation_line_run_data_dict[span.name] = _LineRunData._from_root_span(span)
+                evaluations.append(
+                    {
+                        EvaluationKeyName.DISPLAY_NAME: span.name,
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/entities/_trace.py:252]: avoid this also?

Comment (zhengfeiwang) [reply src/promptflow/promptflow/_sdk/entities/_trace.py:252]: yes, will also update here

File: src/promptflow/promptflow/_sdk/_service/templates/index.html [hunk 6]
```diff
@@ -0,0 +1,12 @@
+<!doctype html>
+<html lang="en">
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/_service/templates/index.html:2]: we can public expose the promptflow.start_trace(), when we have real ui

Comment (zhengfeiwang) [reply src/promptflow/promptflow/_sdk/_service/templates/index.html:2]: how about make this a separate PR? I think we need better CHANGELOG with this public

Comment (wangchao1230) [reply src/promptflow/promptflow/_sdk/_service/templates/index.html:2]: We will expose the function first (so we can change sample to expected function) and anounce in CHANGELOG later.

Comment (zhengfeiwang) [reply src/promptflow/promptflow/_sdk/_service/templates/index.html:2]: as offline synced, exposed but not announce

**Review by wangchao1230 — APPROVED** [review]

**Review by brynn-code — APPROVED** [review]

**Pull request merged**